File: game.erl
-module(game).
-export([start/1]).

start([PlayerFile]) ->
  register(game_id_server, spawn(fun() -> game_id_server(0) end)), % Start the game ID server
  {ok, PlayerInfo} = file:consult(PlayerFile),
  io:format("* Rock, Paper Scissors World Championship *~nStarting game log...~n"),
  lists:foreach(fun({Name, Credits}) ->
    io:format("Player ~p started with ~p credits~n", [Name, Credits])
                end, PlayerInfo),
  MasterPid = self(),
  InitialPlayers = [{Name, 0, Credits} || {Name, Credits} <- PlayerInfo],
  spawn_players(InitialPlayers, MasterPid),
  timer:sleep(200),
  game_loop(InitialPlayers, 0, lists:map(fun({Name, _, _}) -> list_to_atom("player_" ++ atom_to_list(Name)) end, InitialPlayers), InitialPlayers).

% Spawn a process for each player
spawn_players([], _MasterPid) -> ok;
spawn_players([{Name, _, Credits} | Rest], MasterPid) ->
  PlayerName = list_to_atom("player_" ++ atom_to_list(Name)),
  PlayerPid = spawn(player, init, [PlayerName, Credits, MasterPid]),
  register(PlayerName, PlayerPid),
  spawn_players(Rest, MasterPid).

% Game loop
game_loop(Players, GameId, PlayerNames, InitialPlayers) ->
  if length(Players) =:= 1 ->
    [{Winner, _, _}] = Players,
    io:format("~nWe have a winner...~n** Tournament Report **~nPlayers:~n"),
    lists:foreach(
      fun({Name, Used, Remaining}) ->
        io:format("~p: credits used: ~p, credits remaining: ~p~n", [Name, Used, Remaining]),
        display_credits(InitialPlayers, Name)
      end,
      Players),
    io:format("---------~n"),
    io:format("Total games: ~p~n", [GameId]),
    io:format("Winner: ~p~n", [Winner]),
    io:format("See you next year...~n"),
    lists:foreach(fun(Name) ->
      case whereis(Name) of
        undefined -> ok;
        Pid -> Pid ! tournament_over
      end
                  end, PlayerNames),
    halt();
    true ->
      receive
        {request_game, From, To} ->
          case {whereis(From), whereis(To)} of
            {undefined, _} ->
              game_loop(Players, GameId, PlayerNames, InitialPlayers);
            {_, undefined} ->
              game_loop(Players, GameId, PlayerNames, InitialPlayers);
            _ ->
              NewGameId = get_new_game_id(),
              io:format("~n+[~p] new game for ~p -> ~p", [NewGameId, From, To]),
              From ! {game_id, NewGameId, To},
              To ! {game_id, NewGameId, From},
              game_loop(Players, NewGameId, PlayerNames, InitialPlayers)
          end;
        {player_out, From, Move, Name, OpponentMove, Name} ->
          NameStr = atom_to_list(Name),
          PlayerName = string:sub_string(NameStr, 8),
          UpdatedPlayers = lists:filter(fun({N, _, _}) -> N =/= list_to_atom(PlayerName) end, Players),
          UpdatedInitialPlayers = update_initial_players(InitialPlayers, PlayerName),
          io:format("~n- (~p) ~p:~p -> ~p:~p = ~p loses [~p credits left]", [GameId, From, Move, Name, OpponentMove, Name, 0]),
          game_loop(UpdatedPlayers, GameId, PlayerNames, UpdatedInitialPlayers);
        Msg ->
          FormattedString = io_lib:format("~p", [Msg]),
          WithoutBraces = string:replace(string:replace(FormattedString, "{", ""), "}", ""),
          [_, GameIdStr, NameBin, FromBin, ResultBin, OpponentMovebin, MoveBin, UsedCreditsStr, CreditsStr] = re:split(WithoutBraces, ","),
          GameId2 = list_to_integer(binary_to_list(GameIdStr)),
          UsedCredits = list_to_integer(binary_to_list(UsedCreditsStr)),
          Credits = list_to_integer(binary_to_list(CreditsStr)),
          Result = list_to_atom(binary_to_list(ResultBin)),
          Name = list_to_atom(binary_to_list(NameBin)),
          From = list_to_atom(binary_to_list(FromBin)),
          OpponentMove = list_to_atom(binary_to_list(OpponentMovebin)),
          Move = list_to_atom(binary_to_list(MoveBin)),

          case Result of
            lose ->
              io:format("~n$ (~p) ~p:~p -> ~p:~p = ~p loses [~p credits left]", [GameId2, From, Move, Name, OpponentMove, Name, Credits - 1]),
              UpdatedPlayers = update_credits(Players, Name, UsedCredits + 1, Credits - 1),
              game_loop(UpdatedPlayers, GameId2, PlayerNames, InitialPlayers);
            tie ->
              io:format("$ (~p) ~p:~p -> ~p:~p = tie, ", [GameId2, From, Move, Name, OpponentMove]),
              game_loop(Players, GameId2, PlayerNames, InitialPlayers);
            _ ->
              game_loop(Players, GameId2, PlayerNames, InitialPlayers)
          end
      after 1000 ->
        game_loop(Players, GameId, PlayerNames, InitialPlayers)
      end
  end.

% Game ID server loop
game_id_server(CurrentId) ->
  receive
    {get_id, Caller} ->
      NewId = CurrentId + 1,
      Caller ! {game_id, NewId},
      game_id_server(NewId)
  end.

get_new_game_id() ->
  game_id_server ! {get_id, self()},
  receive
    {game_id, NewId} -> NewId
  end.

update_initial_players(InitialPlayers, PlayerName) ->
  PlayerAtom = list_to_atom(PlayerName),
  lists:map(fun({N, Used, Remaining}) ->
    case N of
      PlayerAtom ->
        {N, Remaining, 0};
      _ ->
        {N, Used, Remaining}
    end
            end, InitialPlayers).

update_credits(Players, Name, UsedCredits, Credits) ->
  lists:map(fun({N, Used, Remaining}) ->
    PlayerAtom = list_to_atom("player_" ++ atom_to_list(N)),
    case PlayerAtom of
      Name ->
        {N, UsedCredits, Credits};
      _ ->
        {N, Used, Remaining}
    end
            end, Players).

display_credits(Players, Name1) ->
  lists:foreach(
    fun({Name, Used, Remaining}) ->
      case Name =/= Name1 of
        true ->
          io:format("~p: credits used: ~p, credits remaining: ~p ~n", [Name, Used, Remaining]);
        false ->
          ok
      end
    end, Players).

File: Game.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.CountDownLatch;

public class Game implements Runnable {
    private List<Player> players;
    private int gameId = 1;
    private final Object lock = new Object();
    private final CountDownLatch latch;
    private final List<Thread> playerThreads = new ArrayList<>();

    public Game(List<Player> players, CountDownLatch latch) {
        this.players = players;
        this.latch = latch;
    }

    public void startTournament() {
        System.out.println("** Rock, Paper Scissors World Championship **");
        System.out.println("Starting game log...");
        for (Player player : players) {
            Thread playerThread = new Thread(() -> {
                while (!player.isDisqualified()) {
                    Player opponent;
                    int currentGameId;
                    synchronized (lock) {
                        opponent = getRandomValidOpponent(player);
                        if (opponent == null) {
                            break;
                        }
                        currentGameId = gameId;
                        gameId++;
                        System.out.println("+ [" + currentGameId + "] new game for " + player.getName() + " -> " + opponent.getName());
                    }

                    String move1 = player.getRandomMove();
                    String move2 = opponent.getRandomMove();

                    StringBuilder gameResult = new StringBuilder();
                    gameResult.append("$ (").append(currentGameId).append(") ").append(player.getName()).append(":").append(move1)
                            .append(" -> ").append(opponent.getName()).append(":").append(move2);

                    if (move1.equals(move2)) {
                        gameResult.append(" = tie!");
                        while (move1.equals(move2)) {
                            move1 = player.getRandomMove();
                            move2 = opponent.getRandomMove();
                            gameResult.append(" $ (").append(currentGameId).append(") ").append(player.getName()).append(":").append(move1)
                                    .append(" -> ").append(opponent.getName()).append(":").append(move2);
                        }
                    }

                    synchronized (lock) {
                        if (isPlayerWin(move1, move2)) {
                            gameResult.append(" = ").append(opponent.getName()).append(" loses [")
                                    .append(Math.max(opponent.getCredits() - 1, 0)).append(" credits left]");
                            opponent.decreaseCredits();
                            if (opponent.getCredits() == 0) {
                                opponent.setDisqualified(true);
                                int index = gameResult.indexOf("$");
                                gameResult.replace(index, index + 1, "-");
                                System.out.println(gameResult.toString());
                                break;
                            }
                        } else {
                            gameResult.append(" = ").append(player.getName()).append(" loses [")
                                    .append(Math.max(player.getCredits() - 1, 0)).append(" credits left]");
                            player.decreaseCredits();
                            if (player.getCredits() == 0) {
                                player.setDisqualified(true);
                                int index = gameResult.indexOf("$");
                                gameResult.replace(index, index + 1, "-");
                                System.out.println(gameResult.toString());
                                break;
                            }
                        }
                    }

                    System.out.println(gameResult.toString());
                }
            });
            playerThreads.add(playerThread);
            playerThread.start();
        }

        for (Thread thread : playerThreads) {
            try {
                thread.join();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        System.out.println("\nWe have a winner...");
        printTournamentReport();
        latch.countDown();
    }

    private Player getRandomValidOpponent(Player excludePlayer) {
        List<Player> validOpponents = new ArrayList<>();
        for (Player player : players) {
            if (!player.equals(excludePlayer) && !player.isDisqualified()) {
                validOpponents.add(player);
            }
        }
        if (validOpponents.isEmpty()) {
            return null;
        }
        Random random = new Random();
        return validOpponents.get(random.nextInt(validOpponents.size()));
    }

    private synchronized boolean isPlayerWin(String move1, String move2) {
        return (move1.equals("rock") && move2.equals("scissors"))
                || (move1.equals("scissors") && move2.equals("paper"))
                || (move1.equals("paper") && move2.equals("rock"));
    }

    private synchronized void printTournamentReport() {
        System.out.println("** Tournament Report **");
        System.out.println("Players:");
        for (Player player : players) {
            System.out.println(player.getName() + ": credits used: " + player.getCreditsUsed() +
                    ", credits remaining: " + Math.max(player.getCredits(), 0));
        }
        System.out.println("-----");
        System.out.println("Total games: " + (gameId - 1));
        if (!players.isEmpty()) {
            System.out.println("winner: " + fetchWinner(players));
        }
        System.out.println("See you next year...");
    }
    private String fetchWinner(List<Player> players){
        String winner="";
        int maxCredits=0;
        for(Player player: players){
            if(player.getCredits()>maxCredits){
                winner=player.getName();
                maxCredits=player.getCredits();
            }
        }
        return winner;
    }

    @Override
    public void run() {
        startTournament();
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Please provide the player data file as an argument.");
            return;
        }

        List<Player> players = new ArrayList<>();
        String fileName = args[0];

        try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = br.readLine()) != null) {
                line = line.trim();
                if (line.startsWith("{") && line.endsWith(".")) {
                    line = line.substring(1, line.length() - 2);
                    String[] parts = line.split(",");
                    if (parts.length == 2) {
                        String name = parts[0].trim();
                        int credits = Integer.parseInt(parts[1].trim());
                        players.add(new Player(name, credits));
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        if (players.isEmpty()) {
            System.out.println("No players found in the file.");
            return;
        }

        CountDownLatch latch = new CountDownLatch(1);
        Game game = new Game(players, latch);
        Thread masterThread = new Thread(game);
        masterThread.start();

        try {
            latch.await();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}
